README.md
# vscode-wordcount-cjk

[English Version](./README-en.md)

> **注意：本插件基于张鹏程编写的[Word Count CJK](https://marketplace.visualstudio.com/items?itemName=holmescn.vscode-wordcount-cjk)**
> 发布于此，仅用于测试‘VSCode插件市场的发布功能’及个人私用。
> 
> 添加了一些功能：
> 1. 使用字符串数组来配置`wordcount_cjk.statusBarTooltipTemplate`:
> ```
> "wordcount_cjk.statusBarTooltipTemplate": [
>   "Line 1",
>   "Line 2",
>   "Line 3",
> ],
> ```
> 2. 为`wordcount_cjk.statusBarTooltipTemplate`添加了一些对齐辅助用的转义符号
>   因为之前的版本在windows上显示时没法对齐，TAB对齐没有效果
>   这个对齐功能**仅在以微软雅黑为系统默认字体的Windows系统**上有效，(Mac上用的别的字体，而且用TAB对齐没有问题)
>    - `\\ap`: 对齐点。 使用对齐点来切分行，每一行对应的对齐点是对齐的。
>    - `\\l`: 变长空当。与上面的`\\ap`配合使用，可以实现左/右对齐功能。使用效果如下：
>       ```
>       "wordcount_cjk.statusBarTooltipTemplate": [
>         "|左对齐AA\\l|\\ap|左对齐AA\\l|",
>         "|\\l右对齐BBB|\\ap|\\l右对齐BBB|",
>         "|左对齐C\\l|\\ap|左对齐C\\l|",
>         "|\\l右对齐D|\\ap|\\l右对齐D|",
>         "|左对齐EEE\\l|\\ap|左对齐EEE\\l|"
>       ],
>       ```
>       上面的配置将显示为如下内容:   
>       ![a](doc/img/Tooltip-alignment.png)
> 3. 添加了自定义变量，用于 `wordcount_cjk.statusBarTextTemplate` 和 `wordcount_cjk.statusBarTooltipTemplate` 中的字符串模版，可以使用正则表达式来匹配。
>  ```
>  "wordcount_cjk.customVars": [
>    {
>      "name": "中文字数",
>      "regex": "[\\u4E00-\\u9FA5\\uF900-\\uFA2D]"
>    },
>    {
>      "name": "英文单词数",
>      "regex": "[a-zA-Z_]+"
>    },
>    {
>      "name": "ascii",
>      "regex": "[\\u0000-\\u00FF]"
>    },
>    {
>      "name": "空白",
>      "regex": "\\s"
>    },
>    {
>      "name": "常用标点符号",
>      "regex": "[，。；“”‘’：,.;\"']"
>    },
>    {
>      "name": "非中文非ASCII",
>      "regex": "[^\\u0000-\\u00FF\\u4E00-\\u9FA5\\uF900-\\uFA2D]"
>    },
>    {
>      "name": "空行",
>      "regex": "\\n(?=\\r?\\n)"
>    }
>  ]
>  ```
> 4. 添加调试模式配置`"wordcount_cjk.debug": true,`，启用后可以在“输出”的“wordcount_cjk”子项中，看到所有自定义变量的计算结果。
> 

VSCode 的 word count 都是老外写的，没有一个能统计中文字数的，所以我就写了一个。

## Features

在这个版本里，统计了：

1. 总字符数.
2. 非空白字符数.
3. 英语单词数.
4. 非 ASCII 码字符数.
5. CJK 字符数.

如果你不输入日语和韩语，可以把 CJK 字符数当成中文字符数。

支持如下类型文件中的统计：

1. Markdown 文件.
2. 纯文本文件.

对于其它类型的文件，比如 ReStructuredText，如果也想统计一下，我还提供了一个 `Word Count` 命令。不过，这个命令只会计数一次。如果想一直跟踪文档字数的变化，可以使用 `Word Count Activate`。要关闭计数功能，可以使用 `Word Count Deactivate` 命令。

统计结果会显示在状态栏，把鼠标悬停在状态栏上，可以看到完整的统计结果。在当然的版本中，CJK 字符是混在一起计算的。
因为我不懂日语和韩语，不知道他们应该怎么统计。如果有哪位朋友有这个需求，并且知道要怎么统计的，请在 issue 区留言
告诉我，谢谢赐教！

## Extension Settings

* `wordcount_cjk.statusBarTextTemplate`: 定制状态栏文本.
* `wordcount_cjk.statusBarTooltipTemplate`: 定制状态栏提示.
* `wordcount_cjk.regexWordChar`: 用来判断一个字符是否为英语单词字符的正则表达式.
* `wordcount_cjk.regexASCIIChar`: 用来判断一个字符是否为 ASCII 码字符的正则表达式.
* `wordcount_cjk.regexWhitespaceChar`: 用来判断一个字符是否为空白字符的正则表达式.
* `wordcount_cjk.activateLanguages`: 设置哪些类型的文件会显示字数统计。

在 `wordcount_cjk.statusBarTextTemplate` 和 `wordcount_cjk.statusBarTooltipTemplate`, 可以使用下面这些占位符:

1. `cjk`: CJK 字符数.
2. `ascii`: ASCII 码字符数.
3. `whitespace`: 空白字符数.
4. `en_words`: 英语单词数.
5. `total`: 总字符数.

从 1.1 版开始，格式化字符串可以进行计算了，比如：`共 ${total - ascii} 字`, 其中的表达式，必须是一个有效的 JavaScript 表达式。

### 英语单词字符

就是能够构成一个单词的字符。默认使用 `\w` 来检测，也就是 `[A-Za-z0-9_]`。这是正则表达式的默认值。

### ASCII 字符

在默认设置中, 使用了正则表达式 `[\u0000-\u00FF]` 来检测, 这包含了两个字符集:

1. `0000-007F`: C0 Controls and Basic Latin (ASCII)
2. `0080-00FF`: C1 Controls and Latin-1 Supplement (Extended ASCII)

### 空白字符

默认使用正则表达式的 `\s` 来检测，也就是:
`[ \f\n\r\t\v\u00a0\u1680\u180e\u2000-\u200a\u2028\u2029\u202f\u205f\u3000\ufeff]`. 也就是不只包含了 ASCII 码中的空白字符, 还包含了 Unicode 中的空白字符.

### CJK 字符

**NOTE**: 这个现在是不可配置的。

内建的正则表达式为 `[\u4E00-\u9FA5\uF900-\uFA2D]`, 包含了:

* 4E00-9FFF: CJK Unified Ideographs
* F900-FAFF: CJK Compatibility Ideographs

如果有需求的话，以后可能会把 CJK 三个字符集分开。

## TODO

1. 无

**Enjoy!**


CHANGELOG.md
# Change Log
All notable changes to the "wordcount-cjk" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [1.5.0] - 2020-11-02
### Enhancement
- Add customVars that can be used in `wordcount_cjk.statusBarTextTemplate` and `wordcount_cjk.statusBarTooltipTemplate`
  ```
  "wordcount_cjk.customVars": [
    {
      "name": "中文字数",
      "regex": "[\\u4E00-\\u9FA5\\uF900-\\uFA2D]"
    },
    {
      "name": "英文单词数",
      "regex": "[a-zA-Z_]+"
    },
    {
      "name": "ascii",
      "regex": "[\\u0000-\\u00FF]"
    },
    {
      "name": "空白",
      "regex": "\\s"
    },
    {
      "name": "常用标点符号",
      "regex": "[，。；“”‘’：,.;\"']"
    },
    {
      "name": "非中文非ASCII",
      "regex": "[^\\u0000-\\u00FF\\u4E00-\\u9FA5\\uF900-\\uFA2D]"
    },
    {
      "name": "空行",
      "regex": "\\n(?=\\r?\\n)"
    }
  ]
  ```

## [1.4.0] - 2020-10-28
### Enhancement
- Use string array in configure of `wordcount_cjk.statusBarTooltipTemplate`: 
```
"wordcount_cjk.statusBarTooltipTemplate": [
  "Line 1",
  "Line 2",
  "Line 3",
],
```
- Add some escaped-text for `wordcount_cjk.statusBarTooltipTemplate`
  because Text-alignment in tooltip on windows is wrong  
  This fix is **ONLY for Windows** with default system font is `Microsoft Yahei UI`
   - `\\ap`: alignment point. Use this to split line, Make sure that each piece is the same length.
   - `\\l`: loose. Spaces of variable length. Use with `\\ap` to achieve left/right alignment;
      ```
      "wordcount_cjk.statusBarTooltipTemplate": [
        "|左对齐AA\\l|\\ap|左对齐AA\\l|",
        "|\\l右对齐BBB|\\ap|\\l右对齐BBB|",
        "|左对齐C\\l|\\ap|左对齐C\\l|",
        "|\\l右对齐D|\\ap|\\l右对齐D|",
        "|左对齐EEE\\l|\\ap|左对齐EEE\\l|"
      ],
      ```
      will show like this:   
      ![a](doc/img/Tooltip-alignment.png)

## [1.2.0] - 2018-06-01
### Enhancement
- Add `activateLanguages` options.
- Add `Word Count Activate` and `Word Count Deactivate` command

### Remove
- Remove the activate method by `.editorconfig`.

## [1.1.0] - 2018-01-28
### Enhancement
- Format string supports calculation. [Issue #2](https://github.com/holmescn/vscode-wordcount-cjk/issues/2)

### Fix
- Add a timer to update count, which fix the first issue of delay count.

## [1.0.1] - 2018-01-22
### Added
- Add a new activate method through `.editorconfig`. [Issue #1](https://github.com/holmescn/vscode-wordcount-cjk/issues/1)

## [Unreleased]
- Initial release


src/extension.ts
'use strict';
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { WordCounter } from './WordCounter';
import { WordCountController } from './WordCountController';
import { DebugOutput } from './debugOutput';

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
    const configuration = vscode.workspace.getConfiguration("wordcount_cjk");
    const debug = new DebugOutput(context);
    let counter = new WordCounter(configuration, debug);
    let controller = new WordCountController(configuration, counter, debug);
    let command1 = vscode.commands.registerCommand('extension.wordcount', () => {
        // This command is used to activate the extension when
        // edit non-standard type files.
        controller.update(/* force = */true);
    });
    let command2 = vscode.commands.registerCommand('extension.wordcountActivate', () => {
        // This command is used to activate the extension when
        // edit non-standard type files.
        controller.activate();
    });
    let command3 = vscode.commands.registerCommand('extension.wordcountDeactivate', () => {
        // This command is used to deactivate the extension
        controller.deactivate();
    });

    context.subscriptions.push(command1);
    context.subscriptions.push(command2);
    context.subscriptions.push(command3);
    context.subscriptions.push(controller);
    context.subscriptions.push(debug);
}

// this method is called when your extension is deactivated
export function deactivate() {
}


src/debugOutput.ts
import {
  window,
  Disposable,
  WorkspaceConfiguration,
  workspace,
  OutputChannel,
  ExtensionContext
} from "vscode";

export class DebugOutput {
  private _debugOn!: boolean
  private output?: OutputChannel;
  private context: ExtensionContext;
  private _disposables: Disposable[] = [];

  constructor(context: ExtensionContext) {
    this.context = context;
    workspace.onDidChangeConfiguration((e) => {
      if (e.affectsConfiguration("wordcount_cjk")) {
        console.log("wordcount_cjk changed");
        const configuration = workspace.getConfiguration("wordcount_cjk");
        this.getConfig(configuration);
        this.debugOn();
        console.log(
          `onDidChangeConfiguration: debug = ${this._debugOn}`
        );
      }
    }, this, this._disposables);
    this.getConfig(workspace.getConfiguration("wordcount_cjk"))
    this.debugOn();
  }

  private debugOn() {
    if (this._debugOn) {
      if (this.output === undefined) {
        this.output = window.createOutputChannel('wordcount_cjk');
        this.context.subscriptions.push(this.output);
      }
      this.output.show();
    } else {
      this.output?.clear();
      this.output?.hide();
      this.output?.dispose();
      this.output = undefined;
    }
  }
  private getConfig(configuration: WorkspaceConfiguration) {
    this._debugOn = configuration.get<boolean>(
      "debug",
      false
    );
  }

  public appendLine(val: string) { this.output?.appendLine(val); }
  public append(val: string) { this.output?.append(val); }

  dispose() {
    this.output?.dispose();
    Disposable.from(...this._disposables).dispose();
  }
}


src/WordCountController.ts
// Import the module and reference it with the alias vscode in your code below
import {
    window,
    StatusBarItem,
    StatusBarAlignment,
    Disposable,
    WorkspaceConfiguration,
    workspace,
} from "vscode";
import { DebugOutput } from "./debugOutput";
import pixelWidth from "./string-pixel-width";
import { WordCounter } from "./WordCounter";

function fontWidth(str: string): number {
    return pixelWidth(str, { font: "Microsoft YaHei" });
}
/**
 * The controller of word count
 */
export class WordCountController {
    private debug?: DebugOutput;
    private _wordCounter: WordCounter;
    private _statusBarItem: StatusBarItem;
    private _disposable: Disposable;
    private _delayUpdateTimer: any;
    private _isActive: boolean;
    private _statusBarTextTemplate!: string;
    private _statusBarTooltipTemplate!: string;
    private _activateLanguages!: Array<string>;

    constructor(
        configuration: WorkspaceConfiguration,
        wordCounter: WordCounter,
        debug?: DebugOutput
    ) {
        this._isActive = false;
        this._wordCounter = wordCounter;
        this.getConfig(configuration);
        this._statusBarItem = window.createStatusBarItem(
            StatusBarAlignment.Left
        );
        this.debug = debug;

        // subscribe to selection change and editor activation events
        let subscriptions: Disposable[] = [];
        window.onDidChangeTextEditorSelection(
            (e) => {
                if (this._activateLanguages.findIndex((str) => {
                    return str === e.textEditor.document.languageId
                }) !== -1) {
                    this.debug?.appendLine(`📨onDidChangeTextEditorSelection ${e.textEditor.document.fileName}`);
                    this._onEvent();
                }
            },
            this,
            subscriptions
        );
        window.onDidChangeActiveTextEditor(
            (e) => {
                if (e === undefined)
                    return;
                if (this._activateLanguages.findIndex((str) => {
                    return str === e.document.languageId
                }) !== -1) {
                    this.debug?.appendLine(`📨onDidChangeTextEditorSelection ${e.document.fileName}`);
                    this._onEvent();
                }
            },
            this,
            subscriptions
        );
        workspace.onDidChangeConfiguration(
            (e) => {
                if (e.affectsConfiguration("wordcount_cjk")) {
                    this.debug?.appendLine(`📨onDidChangeConfiguration`);
                    const configuration = workspace.getConfiguration(
                        "wordcount_cjk"
                    );
                    this.getConfig(configuration);
                    this.update(true);
                }
            },
            this,
            subscriptions
        );

        // create a combined disposable from both event subscriptions
        this._disposable = Disposable.from(...subscriptions);
        this.alignMapCal();
        this.update();
    }

    private getConfig(configuration: WorkspaceConfiguration) {
        this._statusBarTextTemplate = configuration.get<string>(
            "statusBarTextTemplate",
            "共 ${cjk} 字"
        );
        let tooltipCfg: string[] | string = configuration.get<Array<string>>(
            "statusBarTooltipTemplate",
            [
                "中文字数：\\l${cjk}\\ap",
                "非 ASCII 字符数：\\l${total - ascii}\\ap",
                "英文单词数：\\l${en_words}\\ap",
                "非空白字符数：\\l${total - whitespace}\\ap",
                "总字符数：\\l${total}\\ap",
            ]
        );
        if (Array.isArray(tooltipCfg)) {
            tooltipCfg = tooltipCfg.join("\n");
        }
        this._statusBarTooltipTemplate = tooltipCfg
            .replace(/\\n/g, "\n")
            .replace(/\\t/g, "\t");
        this._activateLanguages = configuration.get<Array<string>>(
            "activateLanguages",
            ["markdown", "plaintext"]
        );
    }

    public update(force: boolean = false) {
        let editor = window.activeTextEditor;
        if (!editor) {
            this._statusBarItem.hide();
            return;
        }

        let doc = editor.document;

        if (force || this.shouldActivate(doc.languageId) || this._isActive) {
            // Update word count.
            if (editor.selection.isEmpty) {
                let text = doc.getText();
                this._wordCounter.count(text);
            } else {
                const text = editor.document.getText(editor.selection);
                this._wordCounter.count(text);
            }

            // Update the status bar
            try {
                // console.log(
                //   `2 this._statusBarTextTemplate = ${this._statusBarTextTemplate}`
                // );
                this._statusBarItem.text = this._wordCounter.format(
                    this._statusBarTextTemplate
                );
                this._statusBarItem.tooltip = this.align(
                    this._wordCounter.format(this._statusBarTooltipTemplate)
                );
                this._statusBarItem.show();
            } catch (e) {
                window.showErrorMessage(
                    `Something is wrong when update status bar: ${e}`
                );
            }
        } else {
            this._statusBarItem.hide();
        }
    }
    alignMap: Map<number, Array<number>> = new Map<number, Array<number>>();
    alignMissingMap: Map<number, number> = new Map<number, number>();
    readonly spaceMap = new Map<string, number>([
        [" ", 13],
        [" ", 17],
        [" ", 20],
        [" ", 24],
        [" ", 25],
        [" ", 33],
        [" ", 50],
        [" ", 59],
    ]);
    spaceIdxMap: Array<number> = [];
    spaceIdx2StrMap: Array<string> = [];
    getSpace(pixel: number): string {
        // pixel= (37~199) + 100*N 可为精确值，
        pixel += 37;
        let str = "";
        if (pixel >= 237) {
            // [237, ∞]
            const num = Math.floor(pixel / 200);
            str += "　　".repeat(num);
            pixel -= 200 * num;
        } else if (pixel > 137) {
            // [137, 237)
            str += "　";
            pixel -= 100;
        } else {
            // [37, 137)
        }

        const spaceArg = this.alignMap.get(pixel);
        // if(this.alignMissingMap.has(pixel)){
        //   console.log(`diff ${this.alignMissingMap.get(pixel)} pix`)
        // }

        if (spaceArg) {
            for (let i = 0; i < spaceArg.length; i++) {
                const arg = spaceArg[i];
                str += this.spaceIdx2StrMap[i].repeat(arg);
            }
        }
        return str;
    }
    private alignMapCal() {
        let sum;
        for (const item of this.spaceMap) {
            this.spaceIdxMap.push(item[1]);
            this.spaceIdx2StrMap.push(item[0]);
        }
        function calNext(
            alignMap: Map<number, Array<number>>,
            spaceIdxMap: Array<number>,
            index: number,
            presum: number,
            coefficients: Array<number>
        ) {
            let width: number = spaceIdxMap[index];
            let cycle = index;
            const isLast: boolean = spaceIdxMap.length - 1 === index;

            for (let i = 0; i < 100 / width; i++) {
                sum = width * i + presum;
                const coeff = [...coefficients, i];
                // if (isLast) {
                //   if (sum > 100) {
                //     console.log(`${coeff}=${sum}`);
                //   } else {
                //     console.log(`${coeff}=${sum} =>`);
                //   }
                // } else {
                //   console.log(`${coeff}`);
                // }
                if (sum > 200) {
                    continue;
                }
                if (isLast) {
                    if (!alignMap.has(sum)) {
                        // console.log(`\t==>`);
                        alignMap.set(sum, coeff);
                    }
                }
                calNext(alignMap, spaceIdxMap, index + 1, sum, coeff);
            }
        }
        calNext(this.alignMap, this.spaceIdxMap, 0, 0, []);

        let prei = 0,
            nexti = 0,
            avri = 0;
        for (let i = 0; i < 200; i++) {
            if (nexti <= i) {
                for (let j = i + 1; j < 100; j++) {
                    if (this.alignMap.has(j)) {
                        nexti = j;
                        avri = (prei + nexti) / 2;
                        // console.log(`[${prei},${nexti}]`);
                        break;
                    }
                }
            }
            if (this.alignMap.has(i)) {
                // console.log(`${i}: ${this.alignMap.get(i)}`);
                prei = i;
                avri = (prei + nexti) / 2;
            } else {
                if (i < avri) {
                    this.alignMap.set(i, this.alignMap.get(prei)!);
                    // console.log(`missing ${i} => ${prei} [${prei},${nexti}]`);
                    this.alignMissingMap.set(i, i - prei);
                } else {
                    this.alignMap.set(i, this.alignMap.get(nexti)!);
                    // console.log(`missing ${i} => ${nexti} [${prei},${nexti}]`);
                    this.alignMissingMap.set(i, i - nexti);
                }
            }
        }
    }

    private align(str: string): string {
        let ret = "";
        let lines: string[] = str.split("\n");
        // console.log(lines);
        let linesSplits: string[][] = lines.map<string[]>((line) => {
            return line.split(/\\ap/);
        });
        // console.log(linesSplits);
        const maxSplitsNum = linesSplits.reduce<number>(
            (prev, cur, curId, array) => {
                if (cur.length > prev) return cur.length;
                else return prev;
            },
            0
        );
        var letter = String.fromCharCode(8202);
        const spaceLen = fontWidth(letter);

        const lineNum = lines.length;
        let outLines = new Array<string>();
        lines.forEach((v, i) => {
            outLines[i] = "";
        });
        // maxSplitsNum 列
        for (let i = 0; i < maxSplitsNum; i++) {
            // 对每一列对齐

            // 每一列最大长度
            const maxLen = linesSplits.reduce<number>(
                (prev, cur, curId, array) => {
                    if (i >= cur.length) return prev;
                    const str = cur[i].replace(/\\l/g, "");
                    const len = fontWidth(str);
                    // console.log(
                    //   `${i} ${curId} (${str.length} | ${len}/${spaceLen}=${
                    //     len / spaceLen
                    //   }): [${cur[i].replace(/\\l/, '')}]`
                    // );
                    if (len > prev) return len;
                    else return prev;
                },
                0
            );
            // console.log(maxLen);
            // for (const lineSplits of linesSplits) {
            linesSplits.forEach((lineSplits, index) => {
                const split = lineSplits[i];
                if (split === undefined) return;
                const length = fontWidth(
                    split.replace(/\\l/g, "").replace(/\\h/g, "")
                );
                const spaceNum = (maxLen - length) / spaceLen;
                const additionalSpaces = split.split(/\\h/).length - 1;
                let pieces = split.replace(/\\h/g, "").split("\\l");
                // console.log(
                //   `(${maxLen} - ${length})}):${
                //     maxLen - length
                //   } / ${spaceLen} = ${spaceNum}`
                // );
                outLines[index] +=
                    pieces.shift() +
                    // letter.repeat(Math.round(spaceNum) + additionalSpaces);
                    this.getSpace(maxLen - length);
                outLines[index] += pieces.join();
            });
        }
        ret = outLines.join("\n");
        // console.log(ret);
        return ret;
    }

    private shouldActivate(languageId: string): boolean {
        for (var l of this._activateLanguages) {
            if (l === languageId) {
                return true;
            }
        }

        return false;
    }

    private _onEvent() {
        this.update();

        if (this._delayUpdateTimer) {
            clearTimeout(this._delayUpdateTimer);
        }

        this._delayUpdateTimer = setTimeout(() => {
            this._delayUpdateTimer = null;
            this.update();
        }, 500);
    }

    /**
     * This makes the class disposable.
     */
    dispose() {
        this._statusBarItem.dispose();
        this._disposable.dispose();
    }

    activate() {
        this._isActive = true;
        this.update();
    }

    deactivate() {
        this._isActive = false;
        this._statusBarItem.hide();
    }
}


src/string-pixel-width/index.ts
import deburr from "lodash.deburr";
import { widthsMap, WidthsMap } from "./widthsMap";
type Font = "Microsoft YaHei";

interface Settings {
    bold?: boolean;
    font?: Font;
    italic?: boolean;
    size?: number;
    map?: WidthsMap;
}

const settingsDefaults: Settings = { font: "Microsoft YaHei", size: 100 };

export default function getWidth(string: string, settings?: Settings): number {
    const sett: Settings = Object.assign({}, settingsDefaults, settings);
    const font = sett.font!.toLowerCase();
    const size = sett.size!;
    const variant = 0 + (sett.bold ? 1 : 0) + (sett.italic ? 2 : 0);
    const map = sett?.map || widthsMap;
    let totalWidth = 0;
    deburr(string)
        .split("")
        ?.forEach((char) => {
            if (/[\x00-\x1F]/.test(char)) {
                // non-printable character
                return true;
            }
            // use width of 'x' as fallback for unregistered char
            const widths = map[font][char] || [100, 100, 100, 100];
            const width = widths[variant];
            totalWidth += width;
            return true;
        });
    return totalWidth * (size / 100);
}


src/WordCounter.ts
import { WorkspaceConfiguration, workspace, Disposable } from "vscode";
import { DebugOutput } from "./debugOutput";

/**
 * Justify if a char is ASCII character.
 *
 * 0000-007F: C0 Controls and Basic Latin (ASCII)
 * 0080-00FF: C1 Controls and Latin-1 Supplement (Extended ASCII)
 *
 * Reference:
 * http://houfeng0923.iteye.com/blog/1035321 (Chinese)
 * https://en.wikipedia.org/wiki/Basic_Latin_(Unicode_block)
 * https://en.wikipedia.org/wiki/Latin-1_Supplement_(Unicode_block)
 *
 * @todo Current ASCII contains ASCII and Extended ASCII.
 * This sould be configurable in the future.
 */
const regASCII = "[\\u0000-\\u00FF]";

/**
 * Test if a char is a word character.
 *
 * The regular expression \w rule is used.
 * In Javascript it matches any alphanumeric character including the underscore.
 * Equivalent to [A-Za-z0-9_].
 *
 * @todo the word means should be configurable.
 */
const regWord = "\\w+";

/**
 * Test if a char is a white space.
 *
 * The regular expression \s rule is used.
 * In Javascript it matches a single white space character, including space,
 * tab, form feed, line feed.
 * Equivalent to [ \f\n\r\t\v\u00a0\u1680\u180e\u2000-\u200a\u2028\u2029\u202f\u205f\u3000\ufeff]
 */
const regWhitespace = "\\s";

/**
 * Justify if a char is Chinese character.
 *
 * 4E00-9FFF: CJK Unified Ideographs
 * F900-FAFF: CJK Compatibility Ideographs
 *
 * Reference:
 * http://houfeng0923.iteye.com/blog/1035321 (Chinese)
 * https://en.wikipedia.org/wiki/CJK_Unified_Ideographs
 * https://en.wikipedia.org/wiki/CJK_Compatibility_Ideographs
 *
 * @todo Refine to contains only Chinese Chars.
 */
const regChinese = "[\\u4E00-\\u9FA5\\uF900-\\uFA2D]";

// code modify from https://write.qq.com/portal/public/editor/static/js/utils/countWords.js
namespace QiDian {
    const EN_SPACE = "\u0020";
    const CN_SPACE = "\u3000";
    const CR = "\r";
    const LF = "\n";
    const TAB = "\t";

    export function countRawWords(value: string) {
        value = value.replace(/\.\.\./g, "a");

        value = value.trim();

        if (!value) {
            return 0;
        }

        // 空白字符
        const connectWords: string[] = [
            EN_SPACE,
            CN_SPACE,
            CR,
            LF,
            TAB,
            "—",
            "-",
            "\u00A0",
        ];

        let count = 1;
        const initialCode = value.charCodeAt(0);
        let inWord =
            (initialCode >= 33 && initialCode <= 126) || initialCode === 167;
        for (let i = 1; i < value.length; i++) {
            const charCode = value.charCodeAt(i);
            const isChar = (charCode >= 33 && charCode <= 126) || charCode === 167;
            if (inWord && isChar) {
                continue;
            }
            const isConnectWord = connectWords.indexOf(value[i]) >= 0;
            if (isConnectWord) {
                inWord = false;
                continue;
            }
            count++;
            inWord = isChar;
        }

        return count;
    }
}
interface CustomRegexVar {
    name: string;
    regex: string;
    regexFlags?: string;
}

type CountFunc = (text: string) => number;

/**
 * The main class to provide counter functionality.
 */
export class WordCounter {
    private debug?: DebugOutput;
    /** 格式化函数 */
    private _replaceFuncs: { [key: string]: Function };

    private _internalVarsNames!: string[];
    private _internalVarsRegexMap!: Map<string, RegExp>;
    private _internalVarsFuncMap!: Map<string, CountFunc>;
    private _internalVarsNumMap!: Map<string, number>;

    private _customVarsNames!: string[];
    private _customVarsRegexMap!: Map<string, RegExp>;
    private _customVarsNumMap!: Map<string, number>;

    private readonly _regexFormatReplace: RegExp;

    private _disposable: Disposable;

    constructor(configuration: WorkspaceConfiguration, debug?: DebugOutput) {
        this._regexFormatReplace = /\$\{([^}]+)\}/g;
        this._replaceFuncs = {};
        this.getConfig(configuration);
        let subscriptions: Disposable[] = [];
        workspace.onDidChangeConfiguration(
            (e) => {
                if (e.affectsConfiguration("wordcount_cjk")) {
                    console.log("wordcount_cjk changed");
                    const configuration = workspace.getConfiguration("wordcount_cjk");
                    this.getConfig(configuration);
                }
            },
            this,
            subscriptions
        );
        this.debug = debug;
        this._disposable = Disposable.from(...subscriptions);
    }
    private getConfig(configuration: WorkspaceConfiguration) {
        const regexWordChar = configuration.get<string>("regexWordChar", "\\w+");
        const regexASCIIChar = configuration.get<string>(
            "regexASCIIChar",
            "[\\u0000-\\u00FF]"
        );
        const regexWhitespaceChar = configuration.get<string>(
            "regexWhitespaceChar",
            "\\s"
        );

        this._internalVarsNames = [];
        this._internalVarsRegexMap = new Map<string, RegExp>();
        this._internalVarsNumMap = new Map<string, number>();
        this._internalVarsFuncMap = new Map<string, CountFunc>();

        ([
            ["en_words", regexWordChar],
            ["ascii", regexASCIIChar],
            ["whitespace", regexWhitespaceChar],
            ["cjk", regChinese],
            ["total", (text) => text.length],
            ["qd", (text) => QiDian.countRawWords(text)],
        ] as [string, string | CountFunc][]).forEach((pair) => {
            const name = pair[0],
                val = pair[1];

            this._internalVarsNames.push(name);

            if (typeof val === "string") {
                this._internalVarsRegexMap.set(name, new RegExp(val, "g"));
            } else if (typeof val === "function") {
                this._internalVarsFuncMap.set(name, val);
            }
        });

        this._customVarsNames = [];
        this._customVarsRegexMap = new Map<string, RegExp>();
        this._customVarsNumMap = new Map<string, number>();
        configuration.get<CustomRegexVar[]>("customVars", [
            {
                "name": "中文字数",
                "regex": "[\\u4E00-\\u9FA5\\uF900-\\uFA2D]"
            },
            {
                "name": "英文单词数",
                "regex": "[a-zA-Z_]+"
            },
            {
                "name": "ascii",
                "regex": "[\\u0000-\\u00FF]"
            },
            {
                "name": "空白",
                "regex": "\\s"
            },
            {
                "name": "常用标点符号",
                "regex": "[，。；“”‘’：,.;\"']"
            },
            {
                "name": "非中文非ASCII",
                "regex": "[^\\u0000-\\u00FF\\u4E00-\\u9FA5\\uF900-\\uFA2D]"
            },
            {
                "name": "空行",
                "regex": "\\n(?=\\r?\\n)"
            }
        ]).forEach((customRegexVar) => {
            const name = customRegexVar.name;
            let flag;
            if (customRegexVar.regexFlags === undefined) {
                flag = 'g';
            } else {
                flag = customRegexVar.regexFlags;
                if (customRegexVar.regexFlags.match('g') === null) {
                    flag += 'g';
                }
            }
            this._customVarsNames.push(name)
            this._customVarsRegexMap.set(name, new RegExp(customRegexVar.regex, flag));
            this._customVarsNumMap.set(name, 0);
        });
    }

    private _regexVarCount(
        text: string,
        regexMap: Map<string, RegExp>,
        varName: string,
        numMap: Map<string, number>
    ) {
        let num = 0;
        let regexp = regexMap.get(varName);
        if (regexp === undefined) return;
        let result;
        if (regexp.global === false) {
            regexp = new RegExp(regexp.source, regexp.flags + "g");
            regexMap.set(varName, regexp);
        }
        while ((result = regexp.exec(text)) !== null) {
            num++;
        }
        numMap.set(varName, num);
    }

    private _regexVarsCount(
        text: string,
        regexMap: Map<string, RegExp>,
        numMap: Map<string, number>
    ) {
        regexMap.forEach((regexp, key) => {
            this._regexVarCount(text, regexMap, key, numMap);
        });
        if (this.debug) {
            for (const item of numMap) {
                this.debug.appendLine(`${item[0]}: ${item[1]}`);
            }
        }
    }

    private _internalVarsCount(text: string) {
        this._regexVarsCount(
            text,
            this._internalVarsRegexMap,
            this._internalVarsNumMap
        );
        this._internalVarsFuncMap.forEach((func, key) => {
            this._internalVarsNumMap.set(key, func(text));
        });
    }

    private _customVarsCount(text: string) {
        this._regexVarsCount(
            text,
            this._customVarsRegexMap,
            this._customVarsNumMap
        );
    }

    public count(text: string) {
        this.debug?.appendLine("====== count ======");
        this._internalVarsCount(text);
        this._customVarsCount(text);
    }

    public format(fmt: string): string {
        const _this = this;

        return fmt.replace(this._regexFormatReplace, (matches, expr) => {
            if (!_this._replaceFuncs[matches]) {
                _this._replaceFuncs[matches] = _this._compileExpiession(expr);
            }

            const f = _this._replaceFuncs[matches];
            const customVars = new Array();
            for (const item of this._internalVarsNumMap) {
                customVars.push(item[1]);
            }
            for (const item of this._customVarsNumMap) {
                customVars.push(item[1]);
            }
            return f(...customVars);
        });
    }

    private _compileExpiession(expr: string) {
        const f = new Function(
            ...this._internalVarsNames,
            ...this._customVarsNames,
            `return ${expr};`
        );

        try {
            const customVars = new Array();
            for (const item of this._internalVarsNames) {
                customVars.push(0);
            }
            for (const item of this._customVarsNames) {
                customVars.push(0);
            }
            f(...customVars);
        } catch (e) {
            return new Function(
                ...this._internalVarsNames,
                ...this._customVarsNames,
                `return '无效表达式: ${expr}';`
            );
        }

        return f;
    }

    dispose() {
        this._disposable.dispose();
    }
}
